doc/Native APIs.md

```diff
@@ -0,0 +1,417 @@
+  (boolean).
+- `feedback_label`: Feedback message header (string).
+- `messages`: A dict with three string entries: `completed`, `incomplete`, and
+  `max_attempts_reached`. See below for info on each (object).
+- `components`: A list of `student_view_data` output of all immediate child
+  components which implement `student_view_data`. Child components which don't
+  implement `student_view_data` are omitted from the list (array).
+
+#### `messages`
+
+These messages are displayed as feedback, depending on the state of the block.
+
+- `completed`: Message to be shown if the user successfully completes the
+  problem. May contain HTML (string).
+- `incomplete`: Message to be shown if the user did not successfully complete
+  the problem, but still has available attempts left. May contain HTML (string).
+- `max_attempts_reached`: Message to be shown if the user did not complete the
+  problem and has no available attempts left.
+
+### `student_view_user_state`
+
+- `num_attempts`: Number of attempts used so far (integer).
+- `completed`: `true` if the user successfully completed the problem at least
+  once (boolean).
+- `student_results`: A list of user submitted answers with metadata for each
+  child component. See below for more info (array).
+- `components`: A dict of `student_view_user_state` output of all immediate
+  child components which implement `student_view_user_state`. Dict keys are
+  component IDs and the value corresponding to each key is the output of
+  `student_view_user_state`(object). Child components which don't implement
+  `student_view_user_state` are not included in the dict.
+
+#### `student_results`
+
+The `student_results` field is an array of two-element arrays. Each nested array
+has the form of `[component_name, data]`. Only components which the student
+already submitted are present in the array. The structure of the `data` object
+depends on the type of the child component and is described separately for each
+type of child component below.
+
+### Custom Handlers
+
+#### `submit`
+
+The `submit` XBlock JSON handler is used for submitting student input. The
+`POST` data should be a dict where keys are child component names. The values
+depend on the component type and are described separately for each type of child
+component below.
+
+Returns a JSON object which contains these fields:
+
+- `results`: Same as `student_results` described under
+  `student_view_user_state` above.
+- `completed`: Same as `completed` described under `student_view_user_state`
+  above.
+- `message`: One of the `completed`, `incomplete`, or `max_attempts_reached`
+  feedback messages described under `messages` above. Can be `null` if no
+  messages has been defined. May contain HTML (string).
+- `max_attempts`: Same as `max_attempts` under `student_view_data`.
+- `num_attempts`: Same as `num_attempts` under `student_view_user_state`.
+
+
+Step Builder (`step-builder`)
+-----------------------------
+
+### `student_view_data`
```

Comment:
I removed those fields as part of this PR because `active_step` and `num_attempts` are derived from user state and therefore don't work correctly in `student_view_data`. You might have to publish a random unit in your course to force a refresh of the cache.